core/trino-main/src/main/java/io/trino/sql/planner/optimizations/LimitMaxWriterNodesCount.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.sql.planner.optimizations;
+
+import com.google.common.collect.ImmutableList;
+import io.trino.Session;
+import io.trino.cost.TableStatsProvider;
+import io.trino.execution.warnings.WarningCollector;
+import io.trino.operator.RetryPolicy;
+import io.trino.sql.planner.PartitioningHandle;
+import io.trino.sql.planner.PlanNodeIdAllocator;
+import io.trino.sql.planner.SymbolAllocator;
+import io.trino.sql.planner.SystemPartitioningHandle;
+import io.trino.sql.planner.TypeProvider;
+import io.trino.sql.planner.plan.ExchangeNode;
+import io.trino.sql.planner.plan.PlanNode;
+import io.trino.sql.planner.plan.SimplePlanRewriter;
+import io.trino.sql.planner.plan.TableWriterNode;
+
+import java.util.List;
+
+import static com.google.common.base.Preconditions.checkState;
+import static io.trino.SystemSessionProperties.MAX_HASH_PARTITION_COUNT;
+import static io.trino.SystemSessionProperties.MAX_WRITER_NODES_COUNT;
+import static io.trino.SystemSessionProperties.getRetryPolicy;
+import static io.trino.sql.planner.SystemPartitioningHandle.FIXED_ARBITRARY_DISTRIBUTION;
+import static io.trino.sql.planner.SystemPartitioningHandle.FIXED_HASH_DISTRIBUTION;
+import static io.trino.sql.planner.SystemPartitioningHandle.SCALED_WRITER_HASH_DISTRIBUTION;
+import static io.trino.sql.planner.plan.ExchangeNode.Scope.REMOTE;
+import static io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static java.util.Objects.requireNonNull;
+
+public class LimitMaxWriterNodesCount
```

Comment:
Instead of having a separate rule, you can do this directly inside `AddExchange#visitTableWriter`

Reply:
Yes, it could. I do it here because:
1) Most probably in the future we can make it adaptive basing on stats. 
2) There is a rule that is similar in the sense of setting `PartitioningScheme` - `ApplyPreferredTableExecutePartitioning`. This rule is very simple, just basing on configuration toggle and could be part of `AddExchanges` as well.  

This is why I decided to wrap it within another rule.

Reply:
TBH, I don't think we need a separate rule for this since we are just applying the configuration. In `ApplyPreferredTableExecutePartitioning` we are taking a decision based on estimates to use preferred partitioning or not. So, Its kinda makes sense.

Reply:
And in future if we ever make it adaptive based on stats we can always add a new rule and remove from `AddExchanges`. It shouldn't be a big change.

Reply:
WDYT? @sopel39

Reply:
I think the easier approach is preferred. If it's just few lines of code, then `AddExchange#visitTableWriter` is good enough.

Be sure to have proper testing coverage.

Reply:
@gaurav8297 ,  the rule `LimitMaxWriterNodesCount` is now much complicated. Especially, we skip the rule in some cases that is not skipped in `AddExchange` rule. I do not think that we should merge them. It is complicated it seems to be at the beginning.